machine.go

```diff
@@ -264,10 +284,44 @@ func NewMachine(ctx context.Context, cfg Config, opts ...Opt) (*Machine, error)
 	m.machineConfig = cfg.MachineCfg
 	m.Cfg = cfg
 
+	// Use the socketPath (which must be unique to the vm) to generate a unique ID
+	// that doesn't include any path components or other special characters.
+	uniqueID, err := uniqueIDFromSocketPath(cfg.SocketPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to generate unique machine ID")
+	}
+	m.uniqueID = uniqueID
+
 	m.logger.Debug("Called NewMachine()")
 	return m, nil
 }
 
+func uniqueIDFromSocketPath(socketPath string) (string, error) {
+	absPath, err := filepath.Abs(socketPath)
+	if err != nil {
+		return "", errors.Wrapf(err,
+			"failed to get absolute path of socket path %q", socketPath)
+	}
+
+	// we don't assume all parent directories for the socket yet exist, so we evaluate
+	// symlinks in the path one component at a time to handle any ENOENT errors
+	parentDir := "/"
+	for _, pathPart := range strings.Split(filepath.Dir(absPath), string(os.PathSeparator)) {
+		parentDir = filepath.Join(parentDir, pathPart)
+
+		resolvedDir, err := filepath.EvalSymlinks(parentDir)
+		if err != nil && !os.IsNotExist(err) {
+			return "", errors.Wrapf(err,
+				"failed while attempting to resolve any symlinks in socket path %q", socketPath)
+		} else {
```

Comment:
nit: can lose the else

Reply:
I've gotten conflicting opinions on similar situations (whether there should be an `else` after an if statement with a `return`). I'm personally slightly biased towards using `else` just because it's a little more resilient to changes in the future (if the `return` is moved or something like that), but don't have a strong opinion. Let me know what you think.

Reply:
`lint` will complain about it

Reply:
I'd lean toward not having the else.

Reply:
There doesn't appear to be any lint rules defined for the SDK at the moment so I'll leave it as is for now since I kind of disagree with that linting rule. If anyone else feels differently let me know.

Reply:
> I'd lean toward not having the else.

Okay, I don't have a strong opinion so will update to not use it if I'm the only one who wants the else.

Reply:
So, I was going over everything again and just realized this logic was subtly wrong anyways. What we wanted was to set the parent path using the output of `ResolveSymlinks` only if err was nil, but because of the if statement it was still using the output of `ResolveSymlinks` when the path didn't exist.

I fixed it now and it invalidated this concern.